Trigger doing_it_wrong for CPT order queries with HPOS query args

jorgeatorres opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

With HPOS we introduced a few new capabilities [in order querying](https://developer.woocommerce.com/docs/hpos-order-querying-apis/). In particular, when HPOS is the backing datastore for orders, you can create complex queries using `meta_query` and `field_query`.

We have explored adding a compat layer that would allow some of those queries to work with the CPT datastore, but due to the differences in the underlying db structure this is not always possible, and might not be worth the effort given we want to push HPOS over CPT.

This PR introduces a "doing it wrong" warning when either `meta_query` or `field_query` are being used in an order query and the datastore is the CPT one. Currently, those special arguments are silently ignored, producing a result set that can be confusing.

The PR also adds filter `woocommerce_order_data_store_cpt_query_unsupported_args` which can be used to customize which query vars trigger the notice.

Closes #46076.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure **WordPress posts storage** is selected in WC > Settings > Advanced > Features > Order data storage.
   If necessary, run `wp hpos sync` to sync so that the option becomes available.
1. Make sure that `WP_DEBUG` is set on your test instance so that "doing it wrong" messages are actually logged/triggered.
1. Run the following script (line by line) and confirm that after each command is executed the results match the description in the comment above. 
   ```
   wp> $args = [ 'return' => 'ids', ];
   wp> error_reporting( E_ALL ); 
   wp> ini_set( 'display_errors', true );

   // Test 1 -- Produces no notice.
   wp> wc_get_orders( $args );

   // Test 2 -- Produces a "doing it wrong" notice about 'field_query'.
   wp> wc_get_orders( $args + [ 'field_query' => [ 1, 2, 3, ] ] );

   // Test 3 -- Produces a "doing it wrong" notice about 'meta_query'.
   wp> wc_get_orders( $args + [ 'meta_query' => [ 1, 2, 3, ] ] );

   // Test 4 -- Produces a "doing it wrong" notice about both 'field_query' and 'meta_query'.
   wp> wc_get_orders( $args + [ 'field_query' => [ 1, 2, 3, ], 'meta_query' => [ 1, 2, 3, ] ] );
   ```
1. Still in the WP shell, limit notices to `field_query`:
   ```
   add_filter( 'woocommerce_order_data_store_cpt_query_unsupported_args', fn() => [ 'field_query' ] );
   ```
   Repeat the tests 1-4 in step 3 and confirm that only tests 2 and 4 produce a notice.
1. Drop the following snippet somewhere suitable (like wp-content/mu-plugins/) to "simulate" the default CPT order datastore being extended in 3rd party code:
   ```php
   <?php
   add_filter(
       'woocommerce_order_data_store',
       function ( $ds ) {
 	      if ( ! class_exists( 'MyOwnDatastore', false ) ) {
 		      class MyOwnDatastore extends WC_Order_Data_Store_CPT {
 		      }
	      }
   
 		   return new MyOwnDatastore;
       }
   );
   ```
1. Confirm that no notice is triggered when you run the commands in step 3.
1. Remove the PHP snippet introduced in step 5.
1. Change the order data storage back to HPOS in WC > Settings > Advanced > Features.
1. Confirm that no notice is triggered when you run the commands in step 3.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

jorgeatorres committed:
Trigger doing_it_wrong for CPT queries using HPOS-only args

jorgeatorres committed:
Add changelog

naman03malhotra reviewed (commented):
I am yet to test this, but leaving a comment for next week
> it could happen that 3rd party code adds support for, say, meta_query to the CPT datastore, and yet we would be triggering a warning. 

Can we check if such a thing exists in the Wild? Because this seems like a valid usecase for which we'll be showing a warning.

naman03malhotra reviewed (commented):
Hey @jorgeatorres, probably I am missing something but I am not able to trigger the message with the instructions in the PR

- Legacy POST storage is enabled
- Open `wp shell`
- Run `wc_get_orders( array( 'field_query' => array( 'operator' => 'AND'  ), 'meta_query' => array(), 'return' => 'ids' ) );`
- No warning triggered for me

![image](https://github.com/woocommerce/woocommerce/assets/15019298/530a602b-69e6-487f-8aff-ecad5b2fb95c)

jorgeatorres commented:
@naman03malhotra: You're right. WP-CLI is a different environment than the web one, so the error reporting needs to be explicitly set or it'd just use the defaults (which I guess on your machine are not to display certain warnings). You should still have been able to confirm those messages in your error log file, but an alternative way to enable error reporting is to execute the following (in the `wp shell`) before anything else:

```
error_reporting( E_ALL );
ini_set( 'display_errors', true );
[...]
```

Please let me know if the above works and I'll update the testing instructions accordingly.

naman03malhotra reviewed (commented):
Correct, that works, thanks @jorgeatorres. 

>it could happen that 3rd party code adds support for, say, meta_query to the CPT datastore, and yet we would be triggering a warning.

I'd just take a second opinion from the team regarding this, otherwise code change looks good to me.

coreymckrill commented on plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -1003,6 +1003,28 @@ protected function get_wp_query_args( $query_vars ) {
 	 */
 	public function query( $query_vars ) {
+		// Trigger doing_it_wrong() for query vars only supported in HPOS.
+		$hpos_only_query_vars   = array( 'meta_query', 'field_query' );
```
If this array was customizable via filter hook or something, it seems like that would resolve the concern about 3rd party support... ?

jorgeatorres replied:
That's a good idea. I've added such filter. Thanks for the suggestion @coreymckrill!

jorgeatorres commented:
Hey @naman03malhotra! I've implemented [Corey's suggestion](https://github.com/woocommerce/woocommerce/pull/47457#discussion_r1671020845) which should help with 3rd party code in case it (somehow) adds support for the HPOS-only query vars to CPT. I've reworked the testing instructions to take this into account. Please let me know what you think!

naman03malhotra reviewed (commented):
![image](https://github.com/user-attachments/assets/ae5bdde4-0fe5-4020-964e-1ce7a07941c9)


I am getting this fatal thrown when I add that script, in /mu-plugins directory.

naman03malhotra reviewed (commented):
Though, steps, 1-4 as per the instructions you've given are working perfectly.

jorgeatorres commented:
@naman03malhotra: I can't reproduce and can't think of reasons for such error, but I've updated the snippet with a less 'clever' version to see if that helps. Please let me know if it works for you.

naman03malhotra reviewed (approved):
Tested! works as expected, the filter addition is a really nice idea and middle ground. Thanks @jorgeatorres and @coreymckrill

naman03malhotra reviewed (commented):
The exception is being caused by some class name conflict, which is weird, because I don't see it defined anywhere. I changed the file name, to `MyOwnDatastore2`

```php

<?php
add_filter(
    'woocommerce_order_data_store',
    function ( $ds ) {
       if ( ! class_exists( 'MyOwnDatastore2', false ) ) {
 	      class MyOwnDatastore2 extends WC_Order_Data_Store_CPT {
 	      }
       }

 	   return new MyOwnDatastore2;
    }
);

```

markakk commented:
For now what is the correct way to get the orders based on the custom meta value?
My args passed to `wc_get_orders($args)` are:
```php
Array
(
    [paginate] => 1
    [limit] => 25
    [paged] => 1
    [meta_query] => Array
        (
            [relation] => AND
            [0] => Array
                (
                    [key] => _omnivalt_method
                    [value] => Array
                        (
                            [0] => omnivalt
                            [1] => omnivalt_pt
                            [2] => omnivalt_c
                            [3] => omnivalt_cp
                            [4] => omnivalt_pc
                            [5] => omnivalt_pn
                            [6] => omnivalt_ps
                        )
                    [compare] => IN
                )
        )
)
```
Since version WC 9.2.0 I just get the error "Order query argument (meta_query) is not supported on the current order datastore" and see empty page.

coreymckrill commented:
@markakk the `meta_query` arg already didn't work with the "legacy" WordPress posts storage option for orders. If you go to WooCommerce > Settings > Advanced > Features and enable "High-performance order storage", you should be able to use `meta_query`.
